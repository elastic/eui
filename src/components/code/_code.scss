/*
* 1. Size the code against the text its embedded within.
*/
.euiCode {
  @include euiCodeFont;
  font-size: .9em; /* 1 */
  padding: .2em .5em; /* 1 */
  background: $euiCodeBlockBackgroundColor;

  &.euiCode--transparentBackground {
    background: transparent;
  }

  @include euiCodeSyntaxTokens;
}
