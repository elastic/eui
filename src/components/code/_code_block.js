import React, {
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import FocusTrap from 'focus-trap-react';
import hljs from 'highlight.js';

import {
  EuiButtonIcon,
} from '../button';

import {
  EuiOverlayMask,
} from '../overlay_mask';

import { keyCodes } from '../../services';

const fontSizeToClassNameMap = {
  s: 'euiCodeBlock--fontSmall',
  m: 'euiCodeBlock--fontMedium',
  l: 'euiCodeBlock--fontLarge',
};

export const FONT_SIZES = Object.keys(fontSizeToClassNameMap);

const paddingSizeToClassNameMap = {
  none: '',
  s: 'euiCodeBlock--paddingSmall',
  m: 'euiCodeBlock--paddingMedium',
  l: 'euiCodeBlock--paddingLarge',
};

export const PADDING_SIZES = Object.keys(paddingSizeToClassNameMap);

/**
 * This is the base component extended by EuiCode and EuiCodeBlock. These components
 * share the same propTypes definition with EuiCodeBlockImpl.
 */
export class EuiCodeBlockImpl extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFullScreen: false,
    };
  }

  highlight = () => {
    if (this.props.language) {
      hljs.highlightBlock(this.code);

      if (this.codeFullScreen) {
        hljs.highlightBlock(this.codeFullScreen);
      }
    }
  }

  onKeyDown = event => {
    if (event.keyCode === keyCodes.ESCAPE) {
      event.preventDefault();
      event.stopPropagation();
      this.closeFullScreen();
    }
  };

  toggleFullScreen = () => {
    this.setState(prevState => ({
      isFullScreen: !prevState.isFullScreen,
    }));
  };

  closeFullScreen = () => {
    this.setState({
      isFullScreen: false,
    });
  };

  componentDidMount() {
    this.highlight();
  }

  componentDidUpdate() {
    this.highlight();
  }

  render() {
    const {
      inline,
      children,
      className,
      fontSize,
      language,
      overflowHeight,
      paddingSize,
      transparentBackground,
      ...otherProps
    } = this.props;

    const classes = classNames(
      'euiCodeBlock',
      fontSizeToClassNameMap[fontSize],
      paddingSizeToClassNameMap[paddingSize],
      {
        'euiCodeBlock--transparentBackground': transparentBackground,
        'euiCodeBlock--inline': inline,
      },
      className
    );

    const codeClasses = classNames('euiCodeBlock__code', language);

    const optionalStyles = {};

    if (overflowHeight) {
      optionalStyles.height = overflowHeight;
    }

    const codeSnippet = (
      <code
        ref={ref => { this.code = ref; }}
        className={codeClasses}
        {...otherProps}
      >
        {children}
      </code>
    );

    const wrapperProps = {
      className: classes,
      style: optionalStyles
    };

    if (inline) {
      return (
        <span {...wrapperProps}>
          {codeSnippet}
        </span>
      );
    }

    let fullScreenButton;

    if (!inline && overflowHeight) {
      fullScreenButton = (
        <EuiButtonIcon
          className="euiCodeBlock__fullScreenButton"
          size="s"
          onClick={this.toggleFullScreen}
          iconType={this.state.isFullScreen ? 'cross' : 'fullScreen'}
          color="text"
          aria-label={this.state.isFullScreen ? 'Collapse' : 'Expand'}
        />
      );
    }

    let fullScreenDisplay;

    if (this.state.isFullScreen) {
      {/*
        Force fullscreen to use large font and padding.
      */}
      const fullScreenClasses = classNames(
        'euiCodeBlock',
        'euiCodeBlock--fontLarge',
        'euiCodeBlock-paddingLarge',
        'euiCodeBlock-isFullScreen',
      );

      fullScreenDisplay = (
        <FocusTrap
          focusTrapOptions={{
            clickOutsideDeactivates: true,
            initialFocus: () => this.codeFullScreen,
          }}
        >
          <EuiOverlayMask>
            <div className={fullScreenClasses}>
              <pre className="euiCodeBlock__pre">
                <code
                  ref={ref => { this.codeFullScreen = ref; }}
                  className={codeClasses}
                  tabIndex={0}
                  onKeyDown={this.onKeyDown}
                >
                  {children}
                </code>
              </pre>

              {fullScreenButton}
            </div>
          </EuiOverlayMask>
        </FocusTrap>
      );
    }

    return (
      <div {...wrapperProps}>
        <pre className="euiCodeBlock__pre">
          {codeSnippet}
        </pre>

        {/*
          If the below fullScreen code renders, it actually attaches to the body because of
          EuiOverlayMask's React portal usage.
        */}
        {fullScreenButton}
        {fullScreenDisplay}
      </div>
    );
  }
}

EuiCodeBlockImpl.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  paddingSize: PropTypes.oneOf(PADDING_SIZES),

  /**
   * Sets the syntax highlighting for a specific language
   */
  language: PropTypes.string,
  overflowHeight: PropTypes.number,
  fontSize: PropTypes.oneOf(FONT_SIZES),
  transparentBackground: PropTypes.bool,

  /**
   * Displays the passed code in an inline format. Also removes any margins set.
   */
  inline: PropTypes.bool,
};

EuiCodeBlockImpl.defaultProps = {
  transparentBackground: false,
  paddingSize: 'l',
  fontSize: 's',
};
