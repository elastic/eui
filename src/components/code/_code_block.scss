.euiCodeBlock {
  display: block;
  position: relative;
  background: $euiCodeBlockBackgroundColor;
  color: $euiCodeBlockColor;
  position: relative;

  .euiCodeBlock__pre {
    height: 100%;
    overflow: auto;
    display: block;
    white-space: pre-wrap;
  }

  .euiCodeBlock__code {
    display: block;
    line-height: $euiLineHeight;
    font-family: $euiCodeFontFamily;
    font-weight: $euiFontWeightRegular;
  }

  .euiCodeBlock__fullScreenButton {
    position: absolute;
    top: $euiSizeM;
    right: $euiSizeL;
  }

  &.euiCodeBlock-isFullScreen {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    .euiCodeBlock__pre {
      padding: $euiSizeXL !important;
    }
  }

  &.euiCodeBlock--fontSmall {
    font-size: $euiFontSizeXS;
  }

  &.euiCodeBlock--fontMedium {
    font-size: $euiFontSizeS;
  }

  &.euiCodeBlock--fontLarge {
    font-size: $euiFontSize;
  }

  &.euiCodeBlock--paddingSmall {
    .euiCodeBlock__pre {
      padding: $euiSizeS;
    }
  }

  &.euiCodeBlock--paddingMedium {
    .euiCodeBlock__pre {
      padding: $euiSize;
    }
  }

  &.euiCodeBlock--paddingLarge {
    .euiCodeBlock__pre {
      padding: $euiSizeL;
    }
  }

  /**
  ** 1. Size the code against the text its embedded within.
  **/
  &.euiCodeBlock--inline {
    display: inline-block;
    white-space: pre;
    color: $euiTextColor;
    font-size: 90%; // 1
    padding: 0 $euiSizeS;
    background: $euiColorLightestShade;

    .euiCodeBlock__pre {
      padding:  0 $euiSizeXS;
    }

    .euiCodeBlock__code {
      display: inline;
      white-space: normal;
    }
  }

  &.euiCodeBlock--transparentBackground {
    background: transparent;
  }

  .hljs > *::selection {
    // Only change the color if the variable IS a color
    // or else no highlight color shows up at all
    @if type-of($euiCodeBlockSelectedBackgroundColor) == color {
      background-color: $euiCodeBlockSelectedBackgroundColor;
    }
  }

  .hljs-comment,
  .hljs-quote {
    color: $euiCodeBlockCommentColor;
    font-style: italic;
  }

  .hljs-selector-tag {
    color: $euiCodeBlockSelectorTagColor;
    font-weight: bold;
  }

  .hljs-string,
  .hljs-subst,
  .hljs-doctag {
    color: $euiCodeBlockStringColor;
  }

  .hljs-number,
  .hljs-literal,
  .hljs-regexp,
  .hljs-variable,
  .hljs-template-variable,
  .hljs-tag .hljs-attr {
    color: $euiCodeBlockNumberColor;
  }

  .hljs-keyword {
    color: $euiCodeBlockKeywordColor;
    font-weight: bold;
  }

  .hljs-function > .hljs-title {
    color: $euiCodeBlockFunctionTitleColor;
  }

  .hljs-tag {
    color: $euiCodeBlockTagColor;
  }

  .hljs-name {
    color: euiCodeBlockNameColor;
  }

  .hljs-type,
  .hljs-class .hljs-title {
    color: $euiCodeBlockTypeColor;
  }

  .hljs-attribute {
    color: $euiCodeBlockAttributeColor;
  }

  .hljs-symbol,
  .hljs-bullet,
  .hljs-built_in,
  .hljs-builtin-name,
  .hljs-link {
    color: $euiCodeBlockSymbolColor;
  }

  .hljs-params {
    color: $euiCodeBlockParamsColor;
  }

  .hljs-meta {
    color: $euiCodeBlockMetaColor;
  }

  .hljs-title {
    color: $euiCodeBlockTitleColor;
  }

  .hljs-section {
    color: $euiCodeBlockSectionColor;
  }

  .hljs-addition {
    background-color: $euiCodeBlockAdditionBackgroundColor;
    color: $euiCodeBlockAdditionColor;
    display: inline-block;
    width: 100%;
  }

  .hljs-deletion {
    background-color: $euiCodeBlockDeletionBackgroundColor;
    color: $euiCodeBlockDeletionColor;
    display: inline-block;
    width: 100%;
  }

  .hljs-selector-class {
    color: $euiCodeBlockSelectorClassColor;
  }

  .hljs-selector-id {
    color: $euiCodeBlockSelectorIdColor;
  }

  .hljs-emphasis {
    font-style: italic;
  }

  .hljs-strong {
    font-weight: bold;
  }

  .hljs-link {
    text-decoration: underline;
  }
}
