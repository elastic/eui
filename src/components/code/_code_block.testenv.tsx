/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
export const EuiCodeBlockImpl = ({
  children,
  inline,
  'data-test-subj': dataTestSubj,
}: any) => {
  const snippet = <code data-test-subj={dataTestSubj}>{children}</code>;
  return inline ? (
    <span>{snippet}</span>
  ) : (
    <div>
      <pre>{snippet}</pre>
    </div>
  );
};

export const FONT_SIZES: Array<'s' | 'm' | 'l'> = ['s', 'm', 'l'];
export const PADDING_SIZES: Array<'s' | 'm' | 'l' | 'none'> = [
  'none',
  's',
  'm',
  'l',
];
