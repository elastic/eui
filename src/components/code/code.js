import React from 'react';

import {
  EuiCodeBlockImpl,
} from './_code_block';

export const EuiCode = ({
  inline, // eslint-disable-line
  ...rest
}) => {
  return (
    <EuiCodeBlockImpl
      inline={true}
      {...rest}
    />
  );
};

EuiCode.propTypes = {
  ...EuiCodeBlockImpl.propTypes
};
