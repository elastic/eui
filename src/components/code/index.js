export {
  EuiCode,
} from './code';

export {
  EuiCodeBlock,
} from './code_block';
