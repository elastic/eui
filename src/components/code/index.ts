/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export { EuiCode, EuiCodeProps } from './code';
export { EuiCodeBlock, EuiCodeBlockProps } from './code_block';
export { EuiCodeBlockImpl } from './_code_block';
