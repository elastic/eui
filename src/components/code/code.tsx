import { CommonProps } from '../common';

import React, { FunctionComponent, HTMLAttributes } from 'react';

import { EuiCodeBlockImpl } from './_code_block';

export type FontSize = 's' | 'm' | 'l';
export type PaddingSize = 'none' | 's' | 'm' | 'l';

export interface EuiCodeSharedProps {
  paddingSize?: PaddingSize;

  /**
   * Sets the syntax highlighting for a specific language
   * @see http://highlightjs.readthedocs.io/en/latest/css-classes-reference.html#language-names-and-aliases
   * for options
   */
  language?: string;
  overflowHeight?: number;
  fontSize?: FontSize;
  transparentBackground?: boolean;
  isCopyable?: boolean;
}

interface Props extends EuiCodeSharedProps {
  inline?: true;
}

export type EuiCodeProps = CommonProps & Props & HTMLAttributes<HTMLElement>;

export const EuiCode: FunctionComponent<EuiCodeProps> = ({
  inline,
  ...rest
}) => {
  return <EuiCodeBlockImpl inline={true} {...rest} />;
};
