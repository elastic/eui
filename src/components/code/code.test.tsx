/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiCode } from './code';

const code = `var some = 'code';
console.log(some);`;

describe('EuiCode', () => {
  test('renders a code snippet', () => {
    const component = render(<EuiCode {...requiredProps}>{code}</EuiCode>);

    expect(component).toMatchSnapshot();
  });
});
