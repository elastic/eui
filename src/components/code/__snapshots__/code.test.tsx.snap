// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiCode renders a code snippet 1`] = `
<span
  class="euiCodeBlock euiCodeBlock--fontSmall euiCodeBlock--paddingLarge euiCodeBlock--inline prismjs language-none testClass1 testClass2"
>
  <code
    aria-label="aria-label"
    class="euiCodeBlock__code"
    data-test-subj="test subject string"
  >
    var some = 'code';
console.log(some);
  </code>
</span>
`;
